bosonprotocol/boson-protocol-contracts
# Code coverage - protocol contract

This PR adds missing tests to push code coverage towards 100%.

Coverage report: [coverage_345f1f2.zip](https://github.com/bosonprotocol/boson-protocol-contracts/files/9382824/coverage_345f1f2.zip)

- Focus of this PR were facets, bases, clients and libs. There will be another PR addressing only the missing check in `ProtocolDiamond`.
- Mocks are excludes from the coverage test
- `ClientLib.sol` and `ClientBase.sol` are excluded from the coverage test, since no client uses them at the moment

After this PR 100% of statements and lines will be covered. However, branches cannot be 100% covered for the reason explained below. Some of currenly missed branches can be tested with more advanced setup  (might be added later).

- `GroupBase.validateCondition`
  - `if-else` statement goes overe values of `_condition.method` and compare it to enum `EvaluationMethod`. Since all cases are covered, one of ifs will always be true last else path never happens. Code coverage is not smart enough to detect that it is unreachable and reports it as missed branch.
![image](https://user-images.githubusercontent.com/64400885/185598189-a41d881b-f832-4a71-98cd-74e9ebaf8c40.png)

- `ProtocolBase`: modifier `onlyUninitialized` ![image](https://user-images.githubusercontent.com/64400885/185598325-c4e59365-beee-4ce3-8abb-e28394eed370.png)
  -  This modifier is added only to `initialize` methods, which are called during the `facetCut`. `initialize` as such are not part of the interface and therefore cannot be invoked on the diamond the second time to reache the `else` path of `require`
  - It would be possible to test it in one of two ways:
     - make a mock facet that passes clashing `interfaceId` as input to the modifier
     - make a mock facet that has a public function with the modifier and the function is a part of the interface
- ExchangeHandlerFacet
  -  `finalizeExchange`  
  ![image](https://user-images.githubusercontent.com/64400885/185599357-f7250e72-beb3-4cb7-b022-64e39059300d.png)
     - in all cases when `finalizeExchange` is called, functions supply a valid argument
     - To reach `else` path of `require`, a mock function is needed to supply invalid value.
  - `getValidBuyer`
![image](https://user-images.githubusercontent.com/64400885/185599968-4fded469-b039-40b1-a39f-077a4d1e0275.png)
     - `else` branch of `require` is unreachable since we don't have a `deactivateBuyer` function
  - `transferTwins`
     - `else` path is unreachable since one of `twin.tokenType == XX` will always be true. The corresponding check for the amount will then also be true since we are doing `bundleSupplyChecks` when bundle is created. To test this, we'd need a mock function that updates `twin.supplyAvailabe`, so one of the supply checks would fail.
- `FundsLib.releaseFunds` 
![image](https://user-images.githubusercontent.com/64400885/185600655-196b0140-7f69-4e1a-bf31-bc470e21e56b.png)
  - function that call `releaseFunds` always makes sure that the exchange is in one of finalized states.
  - To test it we'd need a mock function that calls `releaseFunds` before exchange is finalized
